bson/default_value_decoders.go

```diff
@@ -1351,14 +1351,17 @@ func decodeDefault(dc DecodeContext, vr ValueReader, val reflect.Value) ([]refle
 		}
 
 		var elem reflect.Value
-		if vDecoder == nil {
+		if vDecoder == nil && idx < val.Len() {
 			elem = val.Index(idx).Elem()
 			switch {
 			case elem.Kind() != reflect.Ptr || elem.IsNil():
 				valueDecoder, err := dc.LookupDecoder(elem.Type())
 				if err != nil {
 					return nil, err
 				}
+				if !elem.CanSet() {
```

Comment:
This change is really nuanced. Can we add a comment here noting why we do this? "If an element is unsettable and allocated, it must be overwritten."

Reply:
I also suggest making the following changes to make this more readable:

```go
isInterfaceSlice := eType.Kind() == reflect.Interface && val.Len() > 0

// If this is not an interface slice with pre-populated elements, we can look
// up the decoder for eType once.
var vDecoder ValueDecoder
if !isInterfaceSlice {
	vDecoder, err = dc.LookupDecoder(eType)
	if err != nil {
		return nil, err
	}
}

for {
	// ...
	if isInterfaceSlice && idx < val.Len() {
		// Decode into an existing interface{} slot.
		
		// ...
	} else {
		// For non-interface slices, or if we've exhuasted the pre-populated
		// slots, we create a fresh value.
		if vDecoder == nil {
			// ...
		}

		//...
	}
}
```